### PR TITLE
extension/cli

### LINKED ISSUE #8: Chosing the caching scenario via CLI
Basic functionality for file-based and hitrate-based caching are rudimentary already implemented in the [`connection`](https://github.com/tfesenbecker/lapis/blob/master/lapis/connection.py) module (see issue #6 for further improvements and restructuring). The option of choosing one of these scenarios should be configurable.

### PR DESCRIPTION
**Pull request by HerrHorizontal**
This is a draft for the simulate CLI. It expands the already existent CLI script to support additional choices in the defintion of the simulator object related to caching.

### CODE REVIEW

**Comment by eileen-kuehn** [comment]
@HerrHorizontal, thanks for this PR. I have one general comment before going into detail. 
Did you check that none of these CLI methods is used in one of the scripts from Tabea? As long as we cannot guarantee I would recommend adding the functions that you require instead of updating the name of arguments. On this way none of the existing scripts would break.

**Comment by HerrHorizontal** [comment]
> @HerrHorizontal, thanks for this PR. I have one general comment before going into detail.
> Did you check that none of these CLI methods is used in one of the scripts from Tabea? As long as we cannot guarantee I would recommend adding the functions that you require instead of updating the name of arguments. On this way none of the existing scripts would break.

So the changes related to this PR did only affect the behavior in the `cli/simulate.py` script, which isn't used in the scripts written by @tfesenbecker. She solely relies on the `custom_simulate.py` script, which except for some cosmetic changes, hasn't changed at all.

Indeed, the `cli/simulate.py` script is meant to replace the `custom_simulate.py` by mimicking the behavior of the latter.

File: lapis/cli/simulate.py [hunk 1]
```diff
@@ -88,34 +102,55 @@ def cli(ctx, seed, until, log_tcp, log_file, log_telegraf, calculation_efficienc
...
 @click.pass_context
-def static(ctx, job_file, pool_file, storage_files, remote_throughput, cache_hitrate):
+def static(ctx, job_file, pre_job_rank, machine_ads, job_ads, scheduler_type, pool_files, storage_files, remote_throughput, filebased_caching, cache_hitrate):
     click.echo("starting static environment")
     simulator = Simulator(seed=ctx.obj["seed"])
-    file, file_type = job_file
+    infile, file_type = job_file
     simulator.create_job_generator(
-        job_input=file,
+        job_input=open(infile, "r"),
```
Comment (maxfischer2781) [inline lapis/cli/simulate.py]: Don't ``open`` the file here (and at similar calls) – ``create_job_generator`` already does that. Note how the original code just passes on ``file`` without opening it.

**Comment by HerrHorizontal** [comment]
It seems to work now, and all tests seem to pass.

File: custom_simulate.py [hunk 2]
```diff
@@ -18,25 +18,28 @@
...
+pre_job_rank_default = "0"
 
-machine_ad_defaults = """
-    requirements = target.requestcpus <= my.cpus
-    rank = 1
-    """.strip()
+# machine_ad_defaults = """
+#     requirements = target.requestcpus <= my.cpus
+#     rank = 1
+#     """.strip()
```
Comment (eileen-kuehn) [inline custom_simulate.py]: Why do you still want to keep those lines?
And is there a reason, that you change the default? The information about the rank is not included in what is defined in the scheduler.

File: custom_simulate.py [hunk 3]
```diff
@@ -18,25 +18,28 @@
...
 
-job_ad_defaults = """
-requirements = my.requestcpus <= target.cpus && my.requestmemory <= target.memory
-rank = 0"""
+# job_ad_defaults = """
+# requirements = my.requestcpus <= target.cpus && my.requestmemory <= target.memory
+# rank = 0"""
```
Comment (eileen-kuehn) [inline custom_simulate.py]: Same as above.

File: lapis/scheduler.py [hunk 4]
```diff
@@ -246,14 +246,13 @@ async def job_finished(self, job):
 class CondorJobScheduler(JobScheduler):
     """
     Goal of the htcondor job scheduler is to have a scheduler that somehow
-    mimics how htcondor does schedule jobs.
-    Htcondor does scheduling based on a priority queue. The priorities itself
+    mimics how htcondor schedules jobs.
```
Comment (eileen-kuehn) [inline lapis/scheduler.py:249]: Changes in this file do not seem to be related to the purpose of this PR.

File: lapis/simulator.py [hunk 5]
```diff
@@ -138,7 +138,7 @@ def create_storage(
     def create_scheduler(self, scheduler_type):
         self.job_scheduler = scheduler_type(job_queue=self.job_queue)
 
-    def create_connection_module(self, remote_throughput, filebased_caching=True):
+    def create_connection_module(self, remote_throughput, filebased_caching=False):
```
Comment (eileen-kuehn) [inline lapis/simulator.py:141]: What is the purpose of changing the default value here from `True` to `False`? Will this impact experiments we do based on the configurations provided by Tabea? In case of yes, we should leave it as is for reproducibility.

File: lapis/cli/simulate.py [hunk 6]
```diff
@@ -88,34 +108,84 @@ def cli(ctx, seed, until, log_tcp, log_file, log_telegraf, calculation_efficienc
...
     ),
     default=(None, None, None),
+    multiple=True,
+    help="Tuple of `(storage_file,storage_content_file,storage_type)`",
+)
+@click.option(
+    "--remote-throughput",
+    "remote_throughput",
+    type=float,
+    default=10,
```
Comment (eileen-kuehn) [inline lapis/cli/simulate.py]: What is the background of selecting `10` as a default here? Please keep in mind that the standardised data type used inside of Lapis is bytes.

File: lapis/cli/simulate.py [hunk 7]
```diff
@@ -131,30 +201,111 @@ def static(ctx, job_file, pool_file, storage_files, remote_throughput, cache_hit
...
+    ),
+    default=(None, None, None),
+    multiple=True,
+    help="Tuple of `(storage_file,storage_content_file,storage_type)`",
+)
+@click.option(
+    "--remote-throughput",
+    "remote_throughput",
+    type=float,
+    default=10,
```
Comment (eileen-kuehn) [inline lapis/cli/simulate.py]: Same as above.

File: lapis/cli/simulate.py [hunk 8]
```diff
@@ -131,30 +201,111 @@ def static(ctx, job_file, pool_file, storage_files, remote_throughput, cache_hit
...
+        if "dummycluster" in pool_file.name:
+            simulator.create_connection_module(float("Inf"))
         simulator.create_pools(
-            pool_input=file,
-            pool_reader=pool_import_mapper[file_type],
+            pool_input=pool_file,
+            pool_reader=pool_import_mapper[pool_file_type],
             pool_type=Pool,
             controller=SimulatedLinearController,
         )
```
Comment (eileen-kuehn) [inline lapis/cli/simulate.py]: This seems to be duplicated code. Please try to avoid this.

File: lapis/cli/simulate.py [hunk 9]
```diff
@@ -168,43 +319,128 @@ def dynamic(ctx, job_file, pool_file):
...
+        job_input=infile,
         job_reader=partial(
             job_import_mapper[file_type],
             calculation_efficiency=ctx.obj["calculation_efficiency"],
         ),
     )
-    simulator.create_scheduler(scheduler_type=CondorJobScheduler)
-    for current_pool in static_pool_file:
-        file, file_type = current_pool
+    click.echo(
```
Comment (eileen-kuehn) [inline lapis/cli/simulate.py]: More duplicated code following here?

**Review by eileen-kuehn — CHANGES_REQUESTED** [review]
Thanks for pushing this forward 💪
One general question: Do you know if one of the scripts of Tabea uses this version of the CLI? If this is the case you should leave attribute names as is and also leave the ordering of attributes as is. Instead you might consider introducing new attributes at the end and also introduce the correct default values to imitate the behaviour we initially had. This at least ensures reproducibility of results and allows to help us checking for breaking/non-breaking changes.
Further, there seems to be a bunch of duplicated code. To improve refactoring and minimise errors on our side, you might consider cleaning up the creation of simulation.

File: custom_simulate.py (hunk 2 above)
Comment (HerrHorizontal) [reply custom_simulate.py]: That's something I have overlooked.

Comment (HerrHorizontal) [reply custom_simulate.py]: I comment this back in.

File: custom_simulate.py (hunk 3 above)
Comment (HerrHorizontal) [reply custom_simulate.py]: Same here

File: lapis/cli/simulate.py (hunk 6 above)
Comment (HerrHorizontal) [reply lapis/cli/simulate.py]: Typo. In `custom_simulate.py` the default is given as `1.0`.

File: lapis/scheduler.py (hunk 4 above)
Comment (HerrHorizontal) [reply lapis/scheduler.py:249]: Do you prefer to have this in a separate PR?

File: lapis/simulator.py (hunk 5 above)
Comment (HerrHorizontal) [reply lapis/simulator.py:141]: Tabea set the default in her scripts to `False`. Since `False` is the new default, to avoid confusion I thought it might be safer to get rid of the additional layer and set it to `False` already here, instead of setting a different default value for the flag in the CLI.

File: lapis/cli/simulate.py (hunk 7 above)
Comment (HerrHorizontal) [reply lapis/cli/simulate.py]: See above.

File: lapis/cli/simulate.py (hunk 8 above)
Comment (HerrHorizontal) [reply lapis/cli/simulate.py]: What is your suggestion? Shall I include this in the `cli` method or write a helper function?

File: lapis/cli/simulate.py (hunk 9 above)
Comment (HerrHorizontal) [reply lapis/cli/simulate.py]: see above

**Comment by HerrHorizontal** [comment]
> Thanks for pushing this forward :muscle:
> One general question: Do you know if one of the scripts of Tabea uses this version of the CLI? If this is the case you should leave attribute names as is and also leave the ordering of attributes as is. Instead you might consider introducing new attributes at the end and also introduce the correct default values to imitate the behaviour we initially had. This at least ensures reproducibility of results and allows to help us checking for breaking/non-breaking changes.

This CLI was not used by Tabea and is never called in her helper scripts. In general, I tried to mimic the behavior of the `custom_simulate.py` script with corresponding default parameter values which was used by her, so in future the CLI can replace her custom scripts.

File: lapis/cli/simulate.py [hunk 10]
```diff
@@ -88,34 +108,84 @@ def cli(ctx, seed, until, log_tcp, log_file, log_telegraf, calculation_efficienc
...
         ),
     )
-    simulator.create_scheduler(scheduler_type=CondorJobScheduler)
+    click.echo(
+        "scheduler configuration: \n "
+        "\tscheduler type: {}\n\n"
+        "\tpre job rank: {} \n\n"
+        "\tmachine classads:\n \t{}\n\n"
+        "\tjob classads: {}".format(scheduler_type, pre_job_rank, machine_ads, job_ads)
+    )
```
Comment (maxfischer2781) [inline lapis/cli/simulate.py]: Use f-strings if you immediately interpolate a string.
```suggestion
    click.echo(
        "scheduler configuration: \n "
        f"\tscheduler type: {scheduler_type}\n\n"
        f"\tpre job rank: {pre_job_rank} \n\n"
        f"\tmachine classads:\n \t{machine_ads}\n\n"
        f"\tjob classads: {job_ads}"
    )
```

File: lapis/cli/simulate.py (hunk 8 above)
Comment (maxfischer2781) [reply lapis/cli/simulate.py]: Try to factor out all common code to a helper function. There seems to be a lot of overlap indeed.

**Review by maxfischer2781 — CHANGES_REQUESTED** [review]
Got some individual suggestions in a few places. Overall, please follow Eileen's request to factor out duplicated code.

File: lapis/cli/simulate.py (hunk 9 above)
Comment (HerrHorizontal) [reply lapis/cli/simulate.py]: I absorbed the code in a helper function.

**Comment by HerrHorizontal** [comment]
> Got some individual suggestions in a few places. Overall, please follow Eileen's request to factor out duplicated code.

Done.

**Comment by HerrHorizontal** [comment]
> Got some individual suggestions in a few places. Overall, please follow Eileen's request to factor out duplicated code.

Done

File: lapis/cli/simulate.py [hunk 11]
```diff
@@ -168,48 +259,167 @@ def dynamic(ctx, job_file, pool_file):
...
+    pre_job_rank,
+    machine_ads,
+    job_ads,
+    scheduler_type,
+    static_pool_files,
+    dynamic_pool_files,
+    storage_files,
+    remote_throughput,
+    filebased_caching,
+    cache_hitrate,
```
Comment (eileen-kuehn) [inline lapis/cli/simulate.py]: Please add proper type hints here.

File: lapis/scheduler.py (hunk 4 above)
Comment (maxfischer2781) [reply lapis/scheduler.py:249]: A separate PR seems more suitable if there are no required changes in the file.

File: lapis/simulator.py (hunk 5 above)
Comment (maxfischer2781) [reply lapis/simulator.py:141]: In the interest of practicality, I think changing a default-that-isn't-relied-on is okay. What do you think @eileen-kuehn?

**Review by maxfischer2781 — COMMENTED** [review]
Thanks for these changes.

Now that the common code has been factored out of the CLI, I am having a hard time finding out what is actually *different* about them. As far as I can tell, ``static``/``dynamic`` uses ``--pool-files`` as ``static_pool_files``/``dynamic_pool_files`` whereas ``hybrid`` has separate ``--static-pool-files`` and ``--dynamic-pool-files``. Is that correct?
If that's the only change, I don't think that's worth having three quasi-equivalent CLIs. Just implement only the ``hybrid`` and people can still get the static/dynamic behaviour by just not providing the other file type.

File: lapis/simulator.py (hunk 5 above)
Comment (eileen-kuehn) [reply lapis/simulator.py:141]: I think as long as nothing relies-on, it's fine. In the long run I tend to completely remove that flag.

**Comment by HerrHorizontal** [comment]
> Thanks for these changes.
> 
> Now that the common code has been factored out of the CLI, I am having a hard time finding out what is actually _different_ about them. As far as I can tell, `static`/`dynamic` uses `--pool-files` as `static_pool_files`/`dynamic_pool_files` whereas `hybrid` has separate `--static-pool-files` and `--dynamic-pool-files`. Is that correct?
> If that's the only change, I don't think that's worth having three quasi-equivalent CLIs. Just implement only the `hybrid` and people can still get the static/dynamic behaviour by just not providing the other file type.

So yes you are right. The only difference is in the pool type and the subsequent creation of the pool in the simulator (the `pool_type` and `controller` differ).

**Review by eileen-kuehn — CHANGES_REQUESTED** [review]
Thanks for your deduplication efforts, this helps a lot.
As @maxfischer2781 already mentioned it might be great to reduce the cli to just one command as the hybrid function manages to set up the whole simulation 👍 
In case you are integrating your `create_simulator` function back into the hybrid function you can also ignore the comment on adding type hints.

File: lapis/cli/simulate.py [hunk 12]
```diff
@@ -24,21 +27,28 @@
...
 
-storage_import_mapper = {"standard": storage_reader}
+storage_import_mapper = {
+    "standard": storage_reader,
+    "filehitrate": storage_reader_filebased_hitrate_caching,
+}
 
-"""Simulation CLI, pay attention to the fact that the random seed is currently set to a
-fixed value"""
+"""Simulation CLI, pay attention to the fact that there is currently only one throughput parameter for all storages available"""
```
Comment (eileen-kuehn) [inline lapis/cli/simulate.py]: You should still add a line break here to make flake8 happy.

**Comment by HerrHorizontal** [comment]
> Thanks for these changes.
> 
> Now that the common code has been factored out of the CLI, I am having a hard time finding out what is actually _different_ about them. As far as I can tell, `static`/`dynamic` uses `--pool-files` as `static_pool_files`/`dynamic_pool_files` whereas `hybrid` has separate `--static-pool-files` and `--dynamic-pool-files`. Is that correct?
> If that's the only change, I don't think that's worth having three quasi-equivalent CLIs. Just implement only the `hybrid` and people can still get the static/dynamic behaviour by just not providing the other file type.

In this case we don't need the group (static, dynamic, hybrid) anymore, aren't we? We could just work with a single command.

**Review by eileen-kuehn — APPROVED** [review]
I don't know if you want another review already. 
From my point of view the code is good to go now 👍

**Review by maxfischer2781 — APPROVED** [review]

**Comment by HerrHorizontal** [comment]
> > Thanks for these changes.
> > Now that the common code has been factored out of the CLI, I am having a hard time finding out what is actually _different_ about them. As far as I can tell, `static`/`dynamic` uses `--pool-files` as `static_pool_files`/`dynamic_pool_files` whereas `hybrid` has separate `--static-pool-files` and `--dynamic-pool-files`. Is that correct?
> > If that's the only change, I don't think that's worth having three quasi-equivalent CLIs. Just implement only the `hybrid` and people can still get the static/dynamic behaviour by just not providing the other file type.
> 
> In this case we don't need the group (static, dynamic, hybrid) anymore, aren't we? We could just work with a single command.

I have further simplified the CLI by getting rid of the unnecessary subcommand. @eileen-kuehn, @maxfischer2781 could you please review?

**Review by eileen-kuehn — APPROVED** [review]
Good to go 👍
